apps/scratch/src/vm/load-crt-project.ts

```diff
@@ -1,11 +1,41 @@
 import VM from "scratch-vm";
 import JSZip from "jszip";
-import { ScratchCrtConfig } from "../types/scratch-vm-custom";
+import { ScratchCrtConfig, ScratchProject } from "../types/scratch-vm-custom";
 import { defaultCrtConfig } from "./default-crt-config";
 
 let nextProject: ArrayBuffer | undefined = undefined;
 let isLoading = false;
 
+/**
+ * Check if all assets listed in project.json exist in the ZIP
+ * @param zip The JSZip instance representing the loaded ZIP file.
+ * @param project The parsed project.json object.
+ * @return A list of missing asset filenames. Empty if all assets are present.
+ */
+const checkAssets = async (
+  zip: JSZip,
+  project: ScratchProject,
+): Promise<string[]> => {
+  const missing: string[] = [];
+
+  for (const target of project.targets) {
+    for (const costume of target.costumes || []) {
+      const filename = `${costume.assetId}.${costume.dataFormat}`;
+      if (!zip.file(filename)) {
+        missing.push(filename);
+      }
+    }
+    for (const sound of target.sounds || []) {
```

Comment:
I would add a linebreak here :)

```suggestion
    }
    
    for (const sound of target.sounds || []) {
```